File: scripts/convert_notebooks_to_literate.jl
"""
    convert_notebooks_to_literate()

Convert Jupyter notebooks in docs/src/exercises to Literate.jl format (.jl files)
so they can be included in the Documenter build and rendered in the documentation.

Literate.jl uses a special format:
- Code cells: normal Julia code
- Markdown cells: `#md # Heading` or `#md Text`
- Output: `#md output` followed by the rendered output
"""

using JSON

function convert_notebooks_to_literate()
    exercises_dir = joinpath(@__DIR__, "..", "docs", "src", "exercises")

    # Find all ex_*.ipynb files
    for (root, dirs, files) in walkdir(exercises_dir)
        for file in files
            if startswith(file, "ex_") && endswith(file, ".ipynb")
                notebook_path = joinpath(root, file)
                exercise_name = splitext(file)[1]  # e.g., "ex_01"
                output_path = joinpath(root, "$(exercise_name).jl")

                println("Converting: $notebook_path → $output_path")
                convert_notebook(notebook_path, output_path)
            end
        end
    end
end

function convert_notebook(notebook_path::String, output_path::String)
    # Read notebook
    nb = JSON.parse(read(notebook_path, String))

    # Convert cells to Literate format
    literate_content = String[]

    for cell in nb["cells"]
        if cell["cell_type"] == "code"
            # Code cell
            source = join(cell["source"], "")

            # Skip Pkg.activate and other setup lines
            lines = split(source, "\n")
            filtered_lines = filter(lines) do line
                !contains(line, "Pkg.activate") &&
                !contains(line, "@__DIR__") &&
                !isempty(strip(line))
            end

            if !isempty(filtered_lines)
                push!(literate_content, join(filtered_lines, "\n"))
                push!(literate_content, "")
            end

        elseif cell["cell_type"] == "markdown"
            # Markdown cell
            source = join(cell["source"], "")

            # Convert to Literate format
            lines = split(source, "\n")
            for line in lines
                if !isempty(strip(line))
                    if startswith(strip(line), "#")
                        # It's a heading - keep the #
                        push!(literate_content, "#md $(line)")
                    elseif startswith(strip(line), "<!--")
                        # Skip HTML comments
                        continue
                    else
                        # Regular text
                        push!(literate_content, "#md $(line)")
                    end
                else
                    push!(literate_content, "#md")
                end
            end
            push!(literate_content, "")
        end
    end

    # Write to file
    open(output_path, "w") do f
        write(f, join(literate_content, "\n"))
    end
end

# Run the conversion
convert_notebooks_to_literate()
println("✓ Conversion complete")

File: scripts/convert_notebooks_to_literate.py
#!/usr/bin/env python3
"""
Convert Jupyter notebooks in docs/src/exercises to Literate.jl format (.jl files)
so they can be included in the Documenter build and rendered in the documentation.
"""

import json
import os
from pathlib import Path

def convert_notebooks_to_literate():
    exercises_dir = Path(__file__).parent.parent / "docs" / "src" / "exercises"

    # Find all ex_*.ipynb files
    for notebook_path in sorted(exercises_dir.glob("*/ex_*.ipynb")):
        exercise_name = notebook_path.stem  # e.g., "ex_01"
        output_path = notebook_path.parent / f"{exercise_name}.jl"

        print(f"Converting: {notebook_path.relative_to(exercises_dir.parent)} → {exercise_name}.jl")
        convert_notebook(notebook_path, output_path)

def convert_notebook(notebook_path, output_path):
    """Convert a Jupyter notebook to Literate.jl format."""

    with open(notebook_path, 'r') as f:
        nb = json.load(f)

    literate_lines = []

    for cell in nb.get("cells", []):
        cell_type = cell.get("cell_type")

        if cell_type == "code":
            # Code cell
            source = "".join(cell.get("source", []))

            # Skip setup lines
            lines = []
            for line in source.split("\n"):
                # Skip Pkg.activate and related setup
                if any(skip in line for skip in ["Pkg.activate", "@__DIR__", "import Pkg"]):
                    continue
                if line.strip():
                    lines.append(line)

            if lines:
                literate_lines.append("\n".join(lines))
                literate_lines.append("")

        elif cell_type == "markdown":
            # Markdown cell
            source = "".join(cell.get("source", []))

            # Convert to Literate format
            for line in source.split("\n"):
                stripped = line.strip()

                # Skip HTML comments
                if stripped.startswith("<!--"):
                    continue

                # Format headings and text
                if stripped.startswith("#"):
                    literate_lines.append(f"#md {line}")
                elif stripped:
                    literate_lines.append(f"#md {line}")
                else:
                    literate_lines.append("#md")

            literate_lines.append("")

    # Write output
    with open(output_path, "w") as f:
        f.write("\n".join(literate_lines))

if __name__ == "__main__":
    convert_notebooks_to_literate()
    print("✓ Conversion complete")
